### ISSUE TITLE
Sleep syscall implementation is buggy

### ISSUE THREAD

**Issue by ronawho** [Qthreads/qthreads#37]
There's a couple of bugs in the sleep, usleep and nanosleep implementations:

* Times passed to the SYS_nanosleep calls in [sleep](https://github.com/Qthreads/qthreads/blob/master/src/syscalls/sleep.c#L43) and [usleep](https://github.com/Qthreads/qthreads/blob/master/src/syscalls/usleep.c#L40) are not being converted to timespec so the syscall fails without sleeping (when syscall interception is enabled.)

* For non-qthreads when HAVE_SYSCALL is undefined or when no supported syscalls were detected, sleep and usleep just return 0. This seems wrong for 2 reasons: they should return the remaining time not slept, but more importantly, it seems like instead of just giving up they should busy wait or yield until the time is up. Maybe something like:

  ```C
  qtimer_t t = qtimer_create();
  qtimer_start(t);
  do {
  #ifdef HAVE_PTHREAD_YIELD
      pthread_yield();
  #elif defined(HAVE_SCHED_YIELD)
      sched_yield();
  #endif
      qtimer_stop(t);
  } while (qtimer_secs(t) < seconds);
  qtimer_destroy(t);
  return 0;
  ```

  This is a high priority bug for us. Even with syscall interception disabled at configure time, somehow qthreads is still intercepting calls to sleep. We've been running into an issue where PMI initialization tries to sleep while waiting for nodes to respond, but the calls were getting intercepted by qthreads and effectively became no-ops because of the above bugs. This has led to cases where PMI init fails because it thinks node are dead because PMI believes it slept/waitied for ~300 seconds for nodes to respond.

**Comment by npe9** [comment]
@ronawho I believe you were right here. I just made a new branch "sleep-fix" which has a test case (always returns true but sleeps enough to eyeball if sleep is working) and the newly timespec'ed code. Could you test it in one of the Chapel programs where PMI was failing?

**Comment by ronawho** [comment]
Yup, I'll try it out. Probably won't get a chance to try it out until tomorrow though.

At a quick glance: 
 -  I'm pretty sure nanosleep is the only one that takes a timespec. For the calls to sleep and usleep I think you still just want to pass seconds, and microseconds respectively 
 - I don't think you want to just return the value of the syscall, think you want to convert the "unslept" time  back to seconds.
 - Just FYI sleep interception seems to occur even if syscall interception is disabled so I'm not sure you want to skip the tests if sycall interception is off
 - I think you still want a final fallback in the non-qthread case in case no syscall support was found, at the very least you don't want to return 0, it should return the amount of time spend unslept. I think the "better" solution is the code in the issue description.
 - usleep is similarly broken

**Comment by ronawho** [comment]
Branch is still broken on crays. `HAVE_SYSCALL` is defined, but `HAVE_DECL_SYS_SLEEP`, `HAVE_DECL_SYS_USLEEP`, and `HAVE_DECL_SYS_NANOSLEEP` are not so we're just hitting one of the `return 0;` branches. 

If I build with `--enable-syscall-interception` then `HAVE_DECL_SYS_NANOSLEEP` is defined to 1, and things seem to work, but we don't build with `--enable-syscall-interception` and I don't think we want to

**Comment by ronawho** [comment]
It must be getting pulled in from somewhere else, but I think you'll also want to include time.h to pull in timespec

**Comment by npe9** [comment]
Aaaah I finally get it. Here's a question, will it break if it calls sleep normally (eg. the qthread is *not* blockable and someone called sleep anyway?)

**Comment by npe9** [comment]
Also what particular program are you running that's failing in PMI? I'd like to be able to reproduce it myself.

**Comment by ronawho** [comment]
>Aaaah I finally get it. Here's a question, will it break if it calls sleep normally (eg. the qthread is not blockable and someone called sleep anyway?)

Not sure what you're asking

>Also what particular program are you running that's failing in PMI? I'd like to be able to reproduce it myself.

Any program can fail. I've been using hello.chpl with CHPL_COMM=gasnet and running with 32 locales. With that I get a failure 1 in ~20 times.  pmi init basically sleeps while it's waiting to hear back from all the nodes. It'll give up after 300 seconds and error out because it thinks some nodes are dead. With a broken sleep what pmi thinks is 300 seconds is really fractions of a fraction of a second.

**Comment by npe9** [comment]
@ronawho The problem as I understand it now is that qthreads shims out the regular sleep implementation which breaks PMI. This behavior makes sense when system call interception is enabled but it doesn't necessarily make sense in stock qthreads, which allows individual qthreads to make system calls AFAIK (@m3m0ryh0l3 is this correct?).

After having a few days to sleep on it, the solution I propose is making a new API function qt_sleep() which would incorporate the current qthreads timer logic and move the interception code to live with the rest of the system call interception functionality. This would solve the PMI problem and making sleeping a qthread an explicit part of the API.

**Comment by m3m0ryh0l3** [comment]
The original intention is that Qthreads should *always* be allowed to make system calls. The distinction between the two ways of handling them (e.g. interception vs not) is really one of accounting: should the qthread making the syscall be treated as if it's still executing (i.e. it occupies a worker thread), or should it be treated as if it's no longer executing and thus the worker thread can do something else?

For most syscalls, this is done by creating a `qt_`-prefixed version (e.g. `qt_select`), and conditionally #def'ing them to override the non-prefixed version (when "interception" is enabled).

Now, sleep was treated differently - it is *always* intercepted - because it's fundamentally just "waiting", rather than doing something complicated, and "sleep()" is often used by novice programmers to mean "do something else for a while". Thus, I decided it was desirable to make it use the yield-loop that you can see in the `sleep.c` file in most cases. That may not have been a good decision! It's worth re-examining that decision, keeping in mind things like "principle of least surprise" - what's the least surprising behavior, from the perspective of a programmer who isn't intimately familiar with the Qthreads internals?

Anyway, naturally, if you shim-out `sleep()`, you have integration challenges with software that isn't always executing in a qthread context. I relied upon `syscall()` to do a "normal" `sleep()` operation in that case. (Qthreads relies on the same trick for all other system calls it "intercepts" as well.)

The original bug here is twofold. First, for `nanosleep()`, the arguments weren't being correctly passed along to the `syscall()` version of nanosleep (aka `syscall(SYS_nanosleep, ...)`). The sleep-fix branch corrects that bug - but it's probably worth using a timer in `test/features/syscalls.c` to ensure that an actual sleep happened (within some delta of acceptable clock inaccuracy).  Second, there's the problem of "what about systems that don't have a working `syscall()` implementation?" I'm honestly stunned that such systems exist and are supported systems, but the approach ronawho suggested is reasonable, in my opinion.

@npe9 I don't exactly understand your question, but I think you're essentially asking whether it makes sense to intercept `sleep()` and `nanosleep()` across the board. I don't think it's necessarily an error or a bug to do it, but rather a design decision. It is, however, orthogonal to what @ronawho was getting at, which is to FIX the implementation.

For what it's worth, in my opinion interception (treating qthreads that make syscalls as if they're no longer executing) is the right thing to do whenever you have a long-lived system call (longer lived than the time it takes to swap a qthread to its own syscall thread), because they are indeed no longer executing (think about long-lived write() operations). The challenge is simply one of testing and portability. That is, however, assuming that we're prizing CPU efficiency over syscall latency (which seems reasonable, given that you're programming with a bajillion tiny threads).

Does that help?

**Comment by npe9** [comment]
@m3m0ryh0l3 Thanks a lot for this.

The essence of my question is really "Is overriding sleep() then yielding the right thing to do or should we allow users to choose between sleep implementations using sleep(3) for the original behavior or qt_sleep() to sleep a qthread specifically".

Did you ever look at Plan 9's  [https://swtch.com/plan9port/man/man3/ioproc.html](ioproc(3)) implementation when you were bringing up Qthreads?

**Comment by m3m0ryh0l3** [comment]
To answer your second question first, no I didn't.

To get back to your first question, I think the question becomes: "What is the behavior that the user/programmer expects to have when he calls sleep()?" We can agree that there are two basic behaviors: first, we stop execution of the underlying pthread for the given timespan (i.e. an OS-level scheduler sleep), or second, we stop execution of the qthread for the given timespan (i.e. a Qthread-level scheduler sleep)?

I think the answer, fundamentally, lies in whether we believe the programmer assumes he is controlling his context or his underlying substrate. That is to say, for instance, if the programmer instructs one thread of a highly parallel multithreaded the program to sleep, is he expecting the CPU core (whichever one his lightweight thread got scheduled on) to remain "occupied", and unavailable for processing any other threads for the duration of that "sleep".

I argue: no, of course not. In a standard pthread program, if I call `sleep()`, I expect the underlying CPU to continue processing other threads, especially if I have more threads than CPUs. When a programmer starts using a highly multithreaded library like Qthreads (or OpenMP or Intel TBB or whatever), that expectation doesn't change. Put another way, I have a very finite expectation of the scope of the effect of the `sleep()` call. When I, as a programmer, call `sleep()`, I expect it to affect THIS thread and no other threads.

Of course, as an implementor of Qthreads and a savvy system software programmer, I'm aware that Qthreads is one of many layers in a stack, and `sleep()`, like any function, affects one particular layer in that stack. I'm also aware that Qthreads is built on top of pthreads, and that by calling pthreads functions directly, I can cause things to happen that violate Qthreads' expectations about its substrate, which will make Qthreads behave badly. For example, I recommend that people stay away from using `pthread_mutex()`'s directly within the context of Qthreads, because the effect of such calls is not limited to the calling Qthread. By the same token, however, I don't attempt to intercept such calls, under the assumption that a good programmer would see the `pthread_` prefix and know that he's doing something funky in a Qthread context. The `sleep()` function, on the other hand, is NOT a pthread-specific library function. Originally, `sleep()` suspended the entire application for the given duration. Pthreads extends that original standard behavioral definition by limiting the scope of effect, rather than maintaining the exact original behavior. It is my contention that Qthreads should do the same: extend the behavioral definition of that function to apply (in a reasonable way) in a new context by further limiting the scope.

**Comment by npe9** [comment]
@m3m0ryh0l3 after thinking about it, I propose we go to an ioproc style model. In the short term we just rename sleep qt_sleep() and move the syscall sleep implementation into the syscall interception code.

In the long term I propose implementing an interface similar to ioproc(3) (qioprocs?), with a FEB acting as the Plan 9 channel and the calls being shunted off to a separate pthread while the communicating thread waits on FEB. Syscall interception in this case is just a matter wrapping the qioprocs(). This *does* require that the user use an explicit api to multiplex I/O onto pthreads, but I suspect this clarity might be good for new projects and system call forwarding can be used to retrofit qthreads onto legacy programs a la massivethreads.

What do you think? @olivier-snl @stelleg?

**Comment by m3m0ryh0l3** [comment]
What is the benefit of this approach? Particularly, why do you think this is better than alternative approaches?

**Comment by ronawho** [comment]
@npe9 I have no idea what Plan9 is or what it does so I don't have any sense of why that might be a good idea or not

Like @m3m0ryh0l3 said, I'm most interested in fixing the buggy code. I'm not against the current behavior of always intercepting sleep. If some third-party library code is called from a qthread, I think you want to intercept sleep. i.e. I'm perfectly happy with the current "always intercept sleep" approach , so long as it actually works correctly.

I think it'd be preferable to fix the sleep implementation instead of just not doing the interception, but at this point I just want something that's not broken. Chapel itself doesn't rely on sleep interception, our module "sleep" turns into chpl_task_sleep, which we implement similar to qthreads sleep, except that we use a schield_yield loop instead of trying to do the syscall for non-qthreads. Longer term though we'll have users interoperating with C libraries which will call sleep and I'll think we'll want sleep interception for that.

And as I mentioned in #43 I think it'd be nice to have a qt_{sleep, usleep, nanosleep} calls, but I think having those is orthogonal to fixing this bug. From a SE standpoint, I think it'd be good to have a qt_{sleep, usleep, nanosleep} that the syscall interception routines and users can call, but it's not my codebase. 

Left to my own devices I would implement a public qt_nansleep, then add a qt_sleep and qt_usleep that just call qt_nanosleep. From there I'd update sleep, usleep, and nanosleep to just call the qt_ equivalent. But if that seems wrong to you, I don't really care what solution you go with.


And @m3m0ryh0l3 
 > I'm honestly stunned that such systems exist and are supported systems, but the approach ronawho suggested is reasonable, in my opinion.

Just to be clear, I'm not aware of any systems that don't support syscalls either. Crays do have syscall support, it's just that we don't build with `--enable-syscall-interception` so qthreads configure ends up skipping the checks for SYS_nanosleep and siblings. We _could_ build with `--enable-syscall-interception`, but I wouldn't want to enable that so close to our release since it impacts more than just sleep interception. It's also listed as "experimental", which never inspired me with much confidence to enable it :)

**Comment by m3m0ryh0l3** [comment]
@ronawho The reason it's always been listed as "experimental" is because I didn't have enough benchmark tests that exercised all those syscalls, AND I hadn't done an analysis to decide if, on average, the intercepted syscalls took long enough to merit offloading them. Certainly in some cases, there's the distinct possibility that it doesn't - like non-blocking `write()` operations - and I didn't create carve-outs for those edge-cases.

**Comment by npe9** [comment]
@ronawho I'm familiarizing myself with the IO subsystem right now. It looks to me like Qthreads already uses something similar to the ioproc model, but instead of having a blocking queue per process there is a queue which uses a pthread pool (haven't looked at in depth enough yet, so @m3m0ryh0l3 can correct me if I'm wrong) rather than manually assigning a call to an ioproc.

Speaking of nanosleep, there already is a qt_nanosleep() in the syscall code so passing sleep calls to that should be trivial. I just need to get a better feel for the code.

**Comment by ronawho** [comment]
For the record and to make sure I understand the commits correctly: 

https://github.com/Qthreads/qthreads/compare/1441082...88291a9 makes it so that sleep, usleep, and nanosleep are no longer intercepted by default. They will only be intercepted if --enable-syscall-interception is thrown at configure time and support for SYS_sleep, SYS_usleep, and SYS_nanosleep respectively are found.

In my experience very few systems (none that I've seen so far) support SYS_sleep or SYS_usleep, which means sleep/usleep probably won't be intercepted. They _might_ end up being intercepted in a round-about way because glibc `sleep()` just fowards to `__nanosleep`, so calling the system sleep will end up forwarding to nanosleep for some systems, which qthreads should be able to intercept. Maybe in the long term, it'd be better to just have sleep/usleep use SYS_nanosleep instead?

Since we don't build with `--enable-syscall-interception` these commits effectively turn off sleep interception for chapel for now. I think that's fine for this release, although we (chapel team) probably want to revisit this decision in the future if we and/or our users end up with 3rd party code that wants to call sleep() and have the worker be able to process other qthreads while the sleep is happening.

**Comment by npe9** [comment]
Yup. That’s pretty much the gist of it. The SYS_sleep and SYS_usleep are rare, but if I understand what was going on in the original implementation it’s just a system call shim interface, so I suspect that doing symbol renaming (eg. –wrap,sleep) at the compiler level might be a better approach. That way we can still rely on the system implementation of the relevant library calls.

**Comment by ronawho** [comment]
https://github.com/chapel-lang/chapel/pull/5446 pulls the commits into our repo and early testing looks fine. 

I think what we have is good enough for this release, but I'm going to leave this issue open because I think the right long term solution is to intercept sleep by default and for that I think the current implementation is insufficient. For that I think you want to avoid using SYS_sleep and SYS_usleep and I think you want a fallback implementation for when syscall support is missing.

**Comment by ronawho** [comment]
Closing, this has been working for us for a while, and if there ever is interest in intercepting sleep again this information is still here.
